android/app-example/src/main/java/com/badoo/ribs/example/auth/AuthDataSource.kt

```diff
@@ -0,0 +1,64 @@
+package com.badoo.ribs.example.auth
+
+import com.badoo.ribs.example.BuildConfig
+import com.badoo.ribs.example.network.UnsplashApi
+import com.badoo.ribs.example.network.model.AccessToken
+import com.jakewharton.rxrelay2.BehaviorRelay
+import io.reactivex.ObservableSource
+import io.reactivex.Single
+
+interface AuthDataSource {
+    val authUpdates: ObservableSource<AuthState>
+    fun login(authCode: String): Single<AccessToken>
+    fun loginAnonymous()
+    fun logout()
+    fun getState(): AuthState
+}
+
+class AuthDataSourceImpl(
+    private val api: UnsplashApi,
+    private val storage: AuthStateStorage,
+    private val accessKey: String = BuildConfig.ACCESS_KEY,
+    private val clientSecret: String = BuildConfig.CLIENT_SECRET
+) : AuthDataSource {
+    private val stateRelay = BehaviorRelay.createDefault<AuthState>(storage.restore())
+
+    override val authUpdates: ObservableSource<AuthState> = stateRelay
+
+    override fun login(authCode: String) =
+        api.requestAccessToken(
+            clientId = accessKey,
+            clientSecret = clientSecret,
+            redirectUri = AuthConfig.redirectUri,
+            code = authCode
+        )
+            .doOnSuccess(::onAuthSuccess)
+
+    override fun loginAnonymous() {
+        val state = AuthState.Anonymous
+        updateState(state)
+    }
+
+    override fun logout() {
+        val state = AuthState.Anonymous
```

Comment:
Shouldn't this be `AuthState.Unauthenticated`?

Reply:
It depends on the behavior we want get. Do we want a user to be redirected to the Welcome screen again or to the feed but as an anonymous user?

Reply:
As a user, I'd personally expect to be thrown to welcome screen. We can always change this as we go.